### ISSUE TITLE
Environment Task Evaluation Shows 0% Success Rate Despite Model Capability

### ISSUE THREAD

**Issue by jorahn** [jorahn/rookworld-rlvr#10]
# Environment Task Evaluation Shows 0% Success Rate Despite Model Capability

## Summary
Environment tasks (A: format) consistently show 0% success rate during GRPO training evaluation, despite the pre-trained RookWorld-LM model demonstrating correct environment task generation in standalone tests. Investigation reveals the model is generating incorrect move sequences during training conditions.

## Environment
- **Model**: jrahn/RookWorld-LM-124M (pre-trained)
- **Framework**: Pure PyTorch implementation, no HuggingFace dependencies  
- **Hardware**: Multi-GPU setup (CUDA:0 for training model, CUDA:1 for reference model)
- **Training**: GRPO with mixed policy/environment tasks (mix_env_ratio=1.0 for testing)

## Expected Behavior
Environment tasks should generate correct state transitions following this format:
```
Prompt: A: rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1+e2e4+
Expected: e2e4+rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1+0.001+0+0
```

## Actual Behavior
During training evaluation, the model generates responses with incorrect move sequences:
```
Prompt: A: rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1+g1h3+  
Generated: e2e4+rnbqkbnr/pppppppp/8/8/4P3/2P4N/PPPP1PPP/RNBQKB1R b KQkq - 1 2+0.001+0+0
```

**Issue**: The model was asked to simulate move `g1h3` but generated move history `e2e4`, resulting in evaluation failure.

### Working Standalone Test
```python
from rookworld_rlvr.model.loader import load_rookworld_model
from rookworld_rlvr.train.policy import CausalLMPolicy, GenerationConfig

model = load_rookworld_model("jrahn/RookWorld-LM-124M")
policy = CausalLMPolicy(model, model, config, device="cpu")

prompt = "A: rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1+e2e4+"
config = GenerationConfig(max_new_tokens=80, temperature=0.5, pad_token_id=tokenizer.eos_token_id)
result = policy.generate_batch([prompt], config)
# Result: Parses successfully with correct move sequence
```

### Failing Training Evaluation
```bash
STEPS=1 MIX_ENV_RATIO=1.0 ./train.sh
# Result: "Environment (A:) Task Performance: Structure Correct: 0.00%"
```

### Fixes Already Applied
1. **EOS Token Handling**: Added `pad_token_id=tokenizer.eos_token_id` to GenerationConfig
2. **Attention Mask**: Fixed `_generate_with_logprobs` to properly handle attention masks during generation
3. **Token Limits**: Increased from 64 to 80 tokens for environment tasks (required: 56-69 tokens)
4. **Evaluator Consistency**: Fixed evaluator to use same token limits as data collector

### Current Status
- ✅ **Pipeline Technical Issues**: Resolved (EOS handling, attention masks, token limits)
- ✅ **Model Capability**: Confirmed - model can generate correct environment responses
- ❌ **Training Consistency**: Model generates wrong moves during training evaluation
- ❌ **Evaluation Success**: Still shows 0% success rate

## Root Cause Hypothesis

The issue appears to be related to **model instruction following** rather than technical pipeline issues. Potential causes:

1. **Context Contamination**: Multi-GPU setup or CUDA context affecting generation consistency
2. **Temperature/Sampling**: Different generation parameters during training vs standalone tests  
3. **Model State**: Training-time model state causing different behavior than eval-time state
4. **Prompt Processing**: Tokenization differences between training and standalone environments

## Impact

- Environment task training data may be learning from incorrect/inconsistent labels
- Mixed task training (policy + environment) cannot be validated properly
- GRPO group-relative advantage computation for environment tasks may be corrupted

## Workaround
Use policy-only training until resolved:
```bash
uv run python train_rookworld_grpo.py --mix-env-ratio 0.0
```

## Files Involved
- `src/rookworld_rlvr/train/evaluator.py:252-339` - Environment task evaluation
- `src/rookworld_rlvr/train/policy.py:73-101, 199-280` - Generation methods
- `src/rookworld_rlvr/data/collector.py:213-322` - Environment data collection

## Additional Context
This issue emerged after implementing comprehensive GRPO training fixes. The pre-trained RookWorld-LM model was specifically trained on environment tasks and should handle them correctly. The fact that standalone tests work but training evaluation fails suggests an environmental or context-dependent issue rather than model incapability.

**Priority**: High - Blocks proper validation of mixed task training, which is critical for GRPO stability.
